### PR TITLE
Loop unroll and instruction interleave for GPU codegen.

### PR DESCRIPTION
**Pull request by JamesTheZ**

### CODE REVIEW

**Review by qiuxiafei — COMMENTED** [review]
Need uint test...

**Review by qiuxiafei — COMMENTED** [review]
Need uint test...

File: tao_compiler/mlir/disc/disc_compiler.cc [hunk 1]
```diff
@@ -425,6 +425,14 @@ LogicalResult LowerHLOToLLVM(ModuleOp m, const DISCLoweringOptions& options) {
   pm.addNestedPass<FuncOp>(
       disc_ral::createDiscUnhandledAtomicRMWConverterPass());
   pm.addNestedPass<FuncOp>(disc_ral::createDiscInputInlineFusionPass());
+  bool mem_intensive_opt_experimental = false;
+  tensorflow::ReadBoolFromEnvVar("DISC_MEM_INTENSIVE_OPT_EXPERIMENTAL", false,
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/disc_compiler.cc:429]: I saw some other PR use this env, maybe we need a global function to check `DISC_MEM_INTENSIVE_OPT_EXPERIMENTAL ` ?

File: tao_compiler/mlir/disc/transforms/codegen_utils.cc [hunk 2]
```diff
@@ -390,5 +392,296 @@ int getReductionTileSizeOnCPU() {
...
+
+  auto lbCstOp = forOp.getLowerBound().getDefiningOp<arith::ConstantIndexOp>();
+  auto ubCstOp = forOp.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+  auto stepCstOp = forOp.getStep().getDefiningOp<arith::ConstantIndexOp>();
+  if (lbCstOp && ubCstOp && stepCstOp) {
+    // Constant loop bounds computation.
+    int64_t lbCst = lbCstOp.value();
+    int64_t ubCst = ubCstOp.value();
+    int64_t stepCst = stepCstOp.value();
+    assert(lbCst >= 0 && ubCst >= 0 && stepCst >= 0 &&
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/codegen_utils.cc:589]: A rookie's question, why only support positive loop bounds, it seems 
`for i = -10 to -1` also works well from this code segment?

File: tao_compiler/mlir/disc/disc_compiler.cc (hunk 1 above)
Comment (qiuxiafei) [reply tao_compiler/mlir/disc/disc_compiler.cc:429]: It's better to let each feature having it's own flag ... It is time to take a look these flags and do some 'feature engineering' ? :P

Comment (JamesTheZ) [reply tao_compiler/mlir/disc/disc_compiler.cc:429]: `tensorflow::ReadBoolFromEnvVar` is already a wrapper. Another wrapper based on that does not save much code.

File: tao_compiler/mlir/disc/transforms/codegen_utils.cc [hunk 3]
```diff
@@ -390,5 +392,296 @@ int getReductionTileSizeOnCPU() {
...
+    }
+
+    int64_t tripCountEvenMultiple = tripCount - (tripCount % unrollFactor);
+    int64_t upperBoundUnrolledCst = lbCst + tripCountEvenMultiple * stepCst;
+    assert(upperBoundUnrolledCst <= ubCst);
+    int64_t stepUnrolledCst = stepCst * unrollFactor;
+
+    // Create constant for 'upperBoundUnrolled' and set epilogue loop flag.
+    generateEpilogueLoop = upperBoundUnrolledCst < ubCst;
+    if (generateEpilogueLoop)
```
[the remark this answers is not in the document]
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/codegen_utils.cc]: Fixed.

**Comment by JamesTheZ** [comment]
> Need uint test...

Done.

File: tao_compiler/mlir/disc/transforms/disc_for_loop_unroll_interleave.cc [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+          // Collect For loop if it is an innermost one.
+          if (!enclosesFloops) result.push_back(floop);
+        }
+      }
+    }
+  }
+  return rootEnclosesFloops;
+}
+
+struct ForLoopUnrollInterleave
```
[the remark this answers is not in the document]
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/disc_for_loop_unroll_interleave.cc:86]: Done. Thanks.

File: tao_compiler/mlir/disc/transforms/codegen_utils.cc (hunk 2 above)
Comment (JamesTheZ) [reply tao_compiler/mlir/disc/transforms/codegen_utils.cc:593]: This is to ease the programming. There is no scenario to optimize loops with negative ub/lb to unroll for optimizing. Only dealing with positive ub/lb is enough for the optimization.

**Review by Yancey1989 — APPROVED** [review]
LGTM

**Pull request merged**